external-testing/README.md

```diff
@@ -0,0 +1,65 @@
+# External Testing
+
+Let's talk about testing process against external resources. What does that
+mean? Well, talking to GitHub API, COPR API, running `fedpkg`, cloning
+repositories...
+
+
+## Use cases
+
+* Interact with an external API (such as GitHub, COPR, Pagure, FAS, k8s)
+  * Authenticated vs. unauthenticated calls.
+* Clone a git repository.
+* Construct a git repository.
+* Run a command to perform an action.
+  * Changes happening only locally.
+  * RO access to a remote service.
+  * R/W access to a remote service.
+* Combination of multiple use cases into a single test case.
+
+
+## Solutions
+
+* flexmock
+
+* pytest fixtures
+
+* requre
+
+* VCRpy
+
+* Betamax
+
+
+## VCRpy vs. requre
+
+| A feature | VCRpy | requre |
+|-----------|-------|--------|
+| Record & replay for python requests | ✓ | ✓ |
+| Match on call stack | ✘ | ✓ |
+| Match on URL & request type | ✓ | ✘ |
+| Calling the same URL twice works as expected | ✓ | ✘ |
+| Record & replay for arbitrary functions | ✘ | ✓ |
+| Filtering sensitive data out of recorded responses | ✓ [1] | ✓ [2] |
+| Easy to use | ✓ [3] | ✘ [4] |
+| Actively maintained | ✘ | ✓ |
+
+* [1] https://vcrpy.readthedocs.io/en/latest/advanced.html#filter-sensitive-data-from-the-request
+* [2] https://requre.readthedocs.io/en/latest/usages/postprocessing.html
+* [3] [You can use a decorator or a context manager](https://vcrpy.readthedocs.io/en/latest/usage.html)
+* [4] [Tenths of lines of code](https://github.com/packit-service/packit/blob/master/tests_recording/__init__.py#L16) [which are very hard to understand](https://github.com/packit-service/packit/blob/master/tests_recording/testbase.py#L22)
+
+### TL;DR
+
+These are my personal (@TomasTomecek) comments and opinions:
+* VCRpy's API and documentation is focused on the record & replay use case
+  which makes it very pleasant to work with.
+* requre's documentation and API is centered around implementation details
+  which makes it very configurable, but unfortunately extremely hard to work
+  with - even to the point that you need to understand internals.
+* VCRpy in action:
+  * [Basic usage](https://github.com/packit-service/packit/pull/785)
+  * [Calling the same URL twice](https://github.com/packit-service/ogr/pull/373)
+* The fact that requre is matching replay via call stack makes it less flexible
```

Comment:
stack info are there important to be able to differenciate to store and replay various functions in generic way. in case focusing just to request module, then it could be simplier, and also there is there is SimpleKeyStorage, what stores just function and his module, and supress all other stack info to be important as an minimal match. But the trouble is, that in that case you have to solve very well which keys to use then (arguments of functions) or wrote own wrappers for specific functuions what will solve internal structure of data. (eg similar to recording function from `git` module what will store Fetch/PushInfo) objects to files